working/terminating-tokens/feature-specification.md

```diff
@@ -0,0 +1,1665 @@
+applies to all of the rules in the middle, transitively.
+
+*   "Declaration" is the context of top level code and class bodies. It's very
+    similar to "statement" except that we can ignore newlines before a couple
+    more tokens there because there's no concept of an expression statement at
+    the top level.
+
+*   "Statement" is the context where statements can appear, in particular
+    expression statements. This is the place where most semicolons appear and
+    where newlines are the most involved.
+
+*   "Expression" is nested inside an expression that must have some explicit
+    closing delimiter. Until that delimiter is reached, we know we haven't
+    reached the end of the expression, so all newlines can be ignored.
+
+The rules are:
+
+*   The initial context is declaration.
+
+*   The `{` and `}` in `block` establish a statement context.
+
+*   The `{` and `}` in `mapLiteral` establish an expression context. (The
+    difference between this and the previous rule is the one reason why the
+    rules can't be defined lexically.)
+
+*   The `(` and `)` in `arguments`, `assertStatement`
+    `forLoopParts`, `formalParameterList`, `primary`, `ifStatement`,
+    `whileStatement`, `doStatement`, and `switchStatement` establish an
+    expression context.
+
+*   The `[` and `]` inside `listLiteral`, `cascadeSelector`, and
+    `unconditionalAssignableSelector` establish an expression context.
+
+*   The `case` and `:` in `switchCase` establish an expression context.
+
+**TODO: Should the first two clauses of `?:` establish expression contexts? I
+haven't found it was needed in the corpus I tested, but it's something we could
+consider.**
+
+When multiple rules surrounding a token establish contexts, the innermost wins.
+So, in:
+
+```dart
+main() {
+  [1, () {
+    a
+    -b
+  }]
+}
+```
+
+The stack of contexts surrounding `-` is:
+
+*   "declaration" (the default)
+*   "statement" (main's block body)
+*   "expression" (the list literal)
+*   "statement" (the lambda body)
+
+The innermost is "statement", so the newline before `-` is significant.
+
+### Terminating tokens
+
+Some languages work by having a set of rules for inserting synthetic semicolon
+tokens when a newline appears in a significant position. The problem with that
+is you have to take extra care to *not* insert them in the many places where a
+semicolon can never appear and where a newline should be ignored.
+
+This proposal takes a slightly different approach. Instead, we say that any
+token may or may not be "terminating". A terminating token is the token *after*
+a newline in a context where that newline could be significant. For example:
+
+```dart
+main() {
+  foo()
+  bar()
+}
+```
+
+Here, `bar` is a terminating token and behaves as if a semicolon appeared before
+it, after the `)` on `foo()`. The "should be significant" is vague. More
+precisely:
+
+*   If a newline character appears after the preceding token's lexeme and
```

Comment:
There's something missing from this paragraph - it feels vacuous.  All newline characters appear between two tokens, no?  Maybe something about "and this token is one of a specific set of tokens"?

Reply:
Yes, but not all pairs of tokens do have newlines between them. :) If there's no newline between a token and the next, then the latter token is not terminating (unless it's `}`).